core/src/com/unciv/logic/CityDistance.kt

```diff
@@ -0,0 +1,141 @@
+        fun compare(a: CityDistance?, b: CityDistance?) : CityDistance? {
+
+            if (a == null && b != null)
+                return b
+            else if (a != null && b == null)
+                return a
+            else if (a == null && b == null)
+                return null
+
+            if (a!!.distance < b!!.distance)
+                return a
+            else if (a.distance > b.distance)
+                return b
+            else {
+                if (a.city.civInfo.isMajorCiv() && b.city.civInfo.isMinorCiv())
+                    return a
+                else if (b.city.civInfo.isMajorCiv() && a.city.civInfo.isMinorCiv())
+                    return b
+            }
+
+            return a
+        }
+    }
+}
+
+/** This class holds information about distance from every tile to the nearest city */
+class CityDistanceData {
+
+    @Transient
+    lateinit var game: GameInfo
+
+    companion object {
+        const val IDENTIFIER_ALL_CIVS = "ALL_CIVS"
+        const val IDENTIFIER_MAJOR_CIVS = "MAJOR_CIVS"
+    }
+
+    private var shouldUpdate: Boolean = true
+
+    /** Identifier -> Map (Tile position -> Distance)
+     *  Identifier is either: Civ name, ALL_CIVS or MAJOR_CIVS */
+    private var data: HashMap<String, HashMap<Vector2, CityDistance?>> = HashMap()
+
+    private fun reset() {
+        data = HashMap()
+        data[IDENTIFIER_ALL_CIVS]= HashMap()
+        data[IDENTIFIER_MAJOR_CIVS] = HashMap()
+    }
+
+    private fun resetPlayer(identifier: String) {
+        data[identifier] = HashMap()
+    }
+
+    private fun updateDistanceIfLower(identifier: String, position: Vector2, city: CityInfo, distance: Int) {
+        val currentDistance = data[identifier]!![position]
+        val newDistance = CityDistance(city, distance)
+        data[identifier]!![position] = CityDistance.compare(currentDistance, newDistance)
+    }
+
+    private fun updateDistances(thisTile: TileInfo, city: CityInfo, owner: CivilizationInfo, isMajor: Boolean) {
+
+        val cityTile = city.getCenterTile()
+        val distance = thisTile.aerialDistanceTo(cityTile)
+
+        updateDistanceIfLower(IDENTIFIER_ALL_CIVS, thisTile.position, city, distance)
+        if (isMajor) {
+            updateDistanceIfLower(IDENTIFIER_MAJOR_CIVS, thisTile.position, city, distance)
+            updateDistanceIfLower(owner.civName, thisTile.position, city, distance)
+        }
+    }
+
+    private fun update() {
+
+        // Clear previous info
+        reset()
+
+        for (player in game.civilizations.asSequence()) {
+
+            // Not interested in defeated players
+            if (player.isDefeated())
+                continue
+
+            val isMajor = player.isMajorCiv()
+            if (isMajor)
+                resetPlayer(player.civName)
+
+            // Update distances for every city and every tile
+            for (city in player.cities.asSequence())
+                for (otherTile in game.tileMap.values.asSequence())
```

Comment:
This goes over the entire map for every city
Sounds to me like a performance point, there are sometimes a hundred cities and ten thousand tiles
We'll need to see how this affects general performance

Reply:
I can think of ways to optimize this

Reply:
The distance recalculation is supposed to be done only when city is founded/razed, so not every turn obviously.

Reply:
In that case it's going to be simply huge
I highly doubt we really need this info for each any every tile in the entire map, it'll be cheaper to calculate when needed

Reply:
This is needed to divide map into tactical zones by cities. Tiles in proximity of the city go to that zone. 
Since city tiles can be bought/expanded/snatched by citadel on each player's turn, we need to update division each player's turn also. We surely don't want to recalculate distances each turn, so that is why they are cached.

Reply:
I understand the tactical zone
But even so, we can agree that e.g. tiles 10 away from the city should not be in the "zone" of a city?
Then already we can massively limit the area we need to initialize

Calculation distance thus limited is also cheap, so we can afford to do it every turn
And not add another huge chunk of data to the save file

Reply:
This is the key piece in the source:
![image](https://user-images.githubusercontent.com/32207817/212522681-4f3ae2cb-15a7-4d5e-a19d-b4d3cbb1ac46.png)

Reply:
If we cut that number to say 7 tiles away it drops even more
So instead of 10,000 tiles * number of cities
It drops to 330 initially
And to 180 for 7-tile radius
That's two orders of magnitude for large maps

Reply:
The logic is the following with the current setup:

1) Tiles in radius 2 around the city belong to the city tactical zone independently of the tile owner.
2) Tiles in radius 3-4 around city:
    a) if owned -> belong to owning city zone
    b) if not owned -> belong to closest city by path zone (in original code, I skipped this part already)
3) Tiles in radius >4 cannot belong to this city's zone

The problems start when there is a tile is on a border between two different civ's cities. 
E.g.:  City A -> * -> * X * <- * <- City B
Which city it should belong to? That is why some tiebreaker needed.

Reply:
You will run into overwriting problem if you will just plainly iterate around cities. Some cities will absorb additional tiles by overwriting wihout any tiebraking.

Reply:
> If we cut that number to say 7 tiles away it drops even more So instead of 10,000 tiles * number of cities It drops to 330 initially And to 180 for 7-tile radius That's two orders of magnitude for large maps

You are gonna run into problem when two closely placed cities with 7 radius zones are going to unevenly divide tiles because on overlap there will be no tiebraking - last city is gonna absorb whole intersection.

Reply:
I'm not saying ditch the "closest tile" logic
I'm saying ditch the "iterate entire map"

Reply:
If it's only applicable for 4 tile radius anyway then that's even better!

Reply:
I just copied original algorithm and tested it - it produces good zones division. If you know how to optimize this computing such that it produces good zones division - you are welcome to suggest.

Reply:
The only idea I have atm: calculate and compare distances only for plots in rings of radius up to 4 around cities, all remaining plot will have  undetermined distances.